xmodule/docs/decisions/0003-library-content-block-schema.rst

```diff
@@ -0,0 +1,201 @@
+
+We will also remove the **mode** field, as it is no longer needed.
+
+The interaction between manual, shuffle, and max_count yields a matrix of 8 different behaviors:
+
++-------------------------------------+---------------------------------------------------------------+-----------------------------------------------------------------+
+|                                     | **manual = False**                                            | **manual = True**                                               |
++-------------------------------------+---------------------------------------------------------------+-----------------------------------------------------------------+
+| **shuffle = True,  max_count = -1** | Entire library in random order (V1's "random mode")           | Manual subset of library in random order                        |
++-------------------------------------+---------------------------------------------------------------+-----------------------------------------------------------------+
+| **shuffle = True,  max_count > 0**  | Random subset of library in random order (V1's "random mode") | Random subset of manual subset of library in random order       |
++-------------------------------------+---------------------------------------------------------------+-----------------------------------------------------------------+
+| **shuffle = False, max_count = -1** | Entire library in original order                              | Manual subset of library in original order ("static mode")      |
++-------------------------------------+---------------------------------------------------------------+-----------------------------------------------------------------+
+| **shuffle = False, max_count > 0**  | Random subset of library in original order                    | Random subset of manual subset of library in original order     |
++-------------------------------------+---------------------------------------------------------------+-----------------------------------------------------------------+
+
+At first, we will only aim to support the "random mode" behaviors plus the new "static mode" behavior. Validation will be used to ensure that the other modes are not available. In the future, we could expect to loosen this restriction.
+
+The final library_content block schema, with all changes, will look like this:
+
+.. list-table::
+   :header-rows: 1
+
+   * - Field Name
+     - Scope, Type
+     - Default
+     - Example
+     - Description
+
+   * - source_library_id
+     - settings, str|None
+     - None
+     - "lib:Open-edX:ExampleLib"
```

Comment:
This is supposed to auto-map v1 library keys to v2 library keys as part of the migration, right? Will that involve actual content data changes or transparent mapping under the covers?

Reply:
Great question. @connorhaugh , do you know what the migration would  do with the keys if we ran it today?

Reply:
@ormsbee T&L's plan is to update the content in place:

* mgmt command: https://github.com/openedx/edx-platform/blob/master/cms/djangoapps/contentstore/management/commands/replace_v1_lib_refs_with_v2_in_courses.py
* the task that it'll spin up for every course in the system: https://github.com/openedx/edx-platform/blob/master/cms/djangoapps/contentstore/tasks.py#L1034

Reply:
that is after they copy every v1 library to a v2 library: https://github.com/openedx/edx-platform/blob/master/cms/djangoapps/contentstore/management/commands/copy_libraries_from_v1_to_v2.py

Reply:
@connorhaugh , @ormsbee and I were talking about the destabilization risk that migrating this field in bulk might cause, especially in this event that a backwards migration is needed. Leaving the V1 keys and place and then mapping V1 keys to V2 libraries at runtime might be safer.

I think we can keep the ADR as it is, but if you're open to it, let's all meet after I'm back on Oct 3 and discuss the migration?

Reply:
Down to talk more about it, but key concerns which motivated the hard cutover choice:
- How do we handle a dynamic mapping of v1-> v2 keys? we'll have to keep that forever, as string matching won't work (libraries can change names).
- We can prove to ourselves that all keys have been mapped using validation commands.
- KISS

Reply:
> How do we handle a dynamic mapping of v1-> v2 keys? we'll have to keep that forever, as string matching won't work (libraries can change names).

For better or worse, Course/Library keys and UsageKeys never really change though. So if it's possible for us to craft the keys of the v2 library content to map 1:1 with the v1 content in way that makes conversion simple, I think that'll be good enough for what we want.

---

For any piece of library content that's being used in a course, the course is effectively using its copy of that content when serving things to the student in the LMS. That should be fine, even if the backing v1 library disappears entirely. So I think we can limit our migration to the point at which a user might check for library updates in Studio.

So say there's a CourseWaffleFlag that gates whether a course is forced to use v2 libraries for "update to latest version" functionality.

If the flag is False, then references to v1 and v2 libraries both work, and you can pull an update from either type of library. We switch based on the library key type.

If the flag is True, then for a given course run, Studio maps the v1 library reference to an equivalent v2 library reference. If that v2 library exists, then doing the update step grabs the latest from that v2 library. If that v2 library does not exist, we say there's no library–we ignore the v1 version of it.

I think that gives us a lot more flexibility for dealing with things that might go wrong with the migration process:

1. It eliminates the need to update a bunch of v1 library references for courses that aren't actively being worked on.
2. We won't distort the meaning of the last published date by mass-publishing every course run.
3. If there are individual errors during the upgrade process, they're less likely to put the course into a broken state–because they'd see those problems come up in their Studio draft and be able to discard their changes. We can then fix those v2 libraries as needed.